documents/development/README.md

```diff
@@ -0,0 +1,539 @@
+
+## Control Structures
+
+- Use `if`, `else if`, and `else` statements to handle multiple conditions.
+- Use `switch` statements for multiple conditions with a default case.
+- Use `while` statements for loops that may not execute.
+- Use `do`/`while` statements for loops that must execute at least once.
+- Use `for` statements for loops with a known number of iterations.
+- The use of `goto` is not allowed, except for error handling.
+- For C, try to prevent deeply nested control structures and prefer early returns or error handling `goto` statements.
+  - To prevent deeply nested control structures, the `CELIX_DO_IF` and `CELIX_GOTO_IF_ERR` macros can also be used.
+
+## Functions and Methods
+
+- Limit functions to a single responsibility or purpose, following the Single Responsibility Principle (SRP).
+- Keep functions short and focused, aiming for a length of fewer than 50 lines.
+- Ensure const correctness.
+- For C functions with a lot of different parameters, consider using an options struct.
+  - An options struct combined with a EMPTY_OPTIONS macro can be used to provide default values and a such
+    options struct can be updated backwards compatible.
+  - An options struct ensure that a lot of parameters can be configured, but also direct set on creation.
+- For C++ functions with a lot of different parameters, consider using a builder pattern.
+  - A builder pattern can be updated backwards compatible.
+  - A builder pattern ensure that a lot of parameters can be configured, but also direct set on construction.
+ 
+## Error Handling and Logging
+
+- For C++, throw an exception when an error occurs.
+- For C, if memory allocation is needed or another error can occur, ensure that a function returns a value 
+  than can be used to check for errors. This can be done by:
+  - Returning a `celix_status_t` and if needed using an out parameter.
+  - Returning a NULL pointer if the function returns a pointer.
+  - Returning a boolean value, where `true` indicates success and `false` indicates failure.
+- Use consistent error handling techniques, such as returning error codes or using designated error handling functions.
+- Log errors, warnings, and other important events using the Apache Celix log helper functions or - for libraries - 
+  the `celix_err` (TBD) functionality. 
+- Always check for errors and log them. 
+- Ensure error handling is correct, using test suite with error injection.
+
+## Error Injection
+
+- Use the Apache Celix error_injector libraries to inject errors in unit tests in a controlled way.
+- Create a separate test suite for error injection tests.
+- Reset error injection setup on the `TearDown` function or destructor of the test fixture.
+- If an - internal or external - function is missing error injection support, add it to the error_injector library.
+  - Try to create small error injector libraries for specific functionality.
```

Comment:
> For error injectors of optional external requirement, we need add `celix_subproject` to make them optional, see `ERROR_INJECTOR_MDNSRESPONDER` for an example.
> 
> For error injectors of optional internal subproject, we need to put it under the protection of subproject's switch. Currently we have no examples in Celix, but I do this in my own projects.
> 
> The above may be too detailed to be put in this document.

Is this really needed? 

I think this can also be done by a  find_package(OPTIONAL). So mdns responder example:

Instead of:
```
celix_subproject(ERROR_INJECTOR_MDNSRESPONDER "Option to enable building the mdnsresponder error injector" OFF)
if (ERROR_INJECTOR_MDNSRESPONDER)
    add_subdirectory(mdnsresponder)
endif ()
```

maybe this:
```
find_package(DNSSD QUIET)
if (DNSSD_FOUND) 
    add_subdirectory(mdnsresponder)
endif ()
```

In latter example if the `RSA_DISCOVERY_ZEROCONF` subproject is enabled, the `find_package(DNSSD REQUIRED)` is called and as result the mdnsresponder error injection library should also automatically build (no extra config needed)

WDYT?

Reply:
Currently, we don't have automatic module dependency deduction within the same package, which I plan to add in a future PR. Note that we do have automatic module dependency deduction between conan packages, which conan gives us for free. Dependency deduction works by manipulating these CMake options in conanfile so that these mocks won't be built if not needed.

In that future PR, I also want to make our error injectors installable so that they can be reused by our downstream users.
For Celix's own optional module's mock, it can reuse the module's option. 
For external library's mock, a separate subproject is needed. 
I found that our error injectors is AI-friendly, and can be generated easily by generative-AI without any IP risks.
Large amounts of reusable error injectors will make our users very happy.

For now we don't need mention anything about this in the coding convention. 
When that future PR is worked out, we can revisit this topic.